{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "639127a1-c7b8-4b85-b228-d8571c20a59e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Answer1\n",
    "\n",
    "#exception is an error that occurs during excution of a program. When that error occurs, \n",
    "#python generate an exception that can be handled, which avoids your program to crash.\n",
    "\n",
    "#syntax error is a error that occurs when python parser detects a syntax error.\n",
    "\n",
    "#Answer2\n",
    "#when exception is not handled , the program will crash and the exception will be printed.\n",
    "\n",
    "try:\n",
    "    a=10\n",
    "    b=0\n",
    "    print(a/b) #exception occurs here\n",
    "finally:\n",
    "    print(\"program over\")\n",
    "#answer3\n",
    "#exceptions are handled by try and except block.\n",
    "\n",
    "\n",
    "try:\n",
    "    a1=10\n",
    "    b1=0\n",
    "    print(a1/b1)\n",
    "except Exception as e:\n",
    "    print(\"exception occured\",e)\n",
    "\n",
    "finally:\n",
    "    print(\"out of the program\")\n",
    "\n",
    "\n",
    "#answer4\n",
    "try:\n",
    "    a2=int(input(\"enter a number\"))\n",
    "    print(a2)\n",
    "\n",
    "except Exception as e:\n",
    "    print(\"Exception occurred:\", e)\n",
    "\n",
    "else:\n",
    "    print(\"no exception occured\")\n",
    "\n",
    "finally:\n",
    "    print(\"out of the program\")\n",
    "\n",
    "#raise is used to throw an exception when a certain condition is not met.\n",
    "age=int(input(\"enter age\"))\n",
    "if age<18:\n",
    "    raise Exception(\"age should be above 18\")\n",
    "\n",
    "\n",
    "#answer5\n",
    "#custom exception is used to create our own exception.\n",
    "age=int(input(\"enter age\"))\n",
    "if age<18:\n",
    "    raise Exception(\"age should be above 18\") #custom exception\n",
    "finally:\n",
    "    print(\"out of the program\")\n",
    "\n",
    "#answer6\n",
    "#exception class is the base class for all the exceptions in python.\n",
    "#exception class is used to handle all the exceptions.\n",
    "\n",
    "password = input(\"Enter password: \")\n",
    "\n",
    "class Pass1Exception(Exception):\n",
    "    def __init__(self, msg):\n",
    "        self.msg = msg\n",
    "\n",
    "def checkpassword(password):\n",
    "   \n",
    "    \n",
    "    if len(password) < 8:\n",
    "        raise Pass1Exception(\"Password is invalid\")\n",
    "        \n",
    "        if password.count(\"@\") + password.count(\"#\") + password.count(\"$\") + password.count(\"&\") < 2:\n",
    "            raise Pass1Exception(\"Password is valid\")\n",
    "\n",
    "try:\n",
    "    checkpassword(password)\n",
    "except Pass1Exception as e:\n",
    "    print(e)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
